draft-ietf-ppm-dap.md

```diff
@@ -1186,27 +1271,157 @@ the last round is reached and each Aggregator computes an output share. All
 told there there are `ceil((ROUNDS+1)/2)` HTTP requests sent, where `ROUNDS` is
 the number of rounds specified by the VDAF.
 
-[OPEN ISSUE: For the moment we don't have an example of a VDAF that requires
-more than two rounds. Eventually we may want to specialize the spec to
-accommodate 1-2 rounds (or even just 1).]
+The state machine for VDAF preparation is shown in {{vdaf-prep-state-machine}}.
+
+~~~ state
+ START --> CONTINUED(round, host_prep_state)
+  |                |                    |
+  |                |                    v
+  +--> REJECTED <--+     FINISHED(out_share)
+~~~
+{: #vdaf-prep-state-machine title="State machine for VDAF preparation."}
+
+State transitions are made when the state is acted upon by the host's local
+inputs and/or a prep step sent by the peer. The initial state is `START`.
+
+~~~ transition
+State START
+
+Input
+
+  vdaf_verify_key: VDAF verification key associated with the task
+  is_leader: boolean indicating whether the host is the Leader
+  agg_param: aggregation parameter for the aggregation job
+  report_id: report ID
+  public_share: report public share
+  host_input_share: host's input share for the report
+
+def VdafPrepInit(vdaf_verify_key,
+                 is_leader,
+                 agg_param,
+                 report_id,
+                 public_share,
+                 host_input_share):
+  try:
+    host_prep_state = Vdaf.prep_init(
+      vdaf_verify_key,
+      0 if is_leader else 1,
+      agg_param,
+      report_id,
+      public_share,
+      host_input_share,
+    )
```

Comment:
This method would then need to call `prep_next(host_prep_state, None)` as is currently done during initialization: https://www.ietf.org/archive/id/draft-ietf-ppm-dap-04.html#name-input-share-preparation

Reply:
I think the intent might be to move this `prep_next` call to https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/461/files#diff-8dbbcb0d121b441734a70cc867271480731a494b50eefdd758c249fdfb5ab84eR1347-R1350 -- if so, this works out for Helper initialization, but would leave the Leader needing to make an additional `prep_next` call after `VdafPrepInit` but before sending its aggregation-initialization message to the Helper.

That's not ideal since users of the rest of the document now need to care about both the "wrapper" interface and the normal VDAF interface; that is, the encapsulation of the VDAF interface is leaking.

Unfortunately, if we do specify `VdafPrepInit` as including the initial `prep_next` call, this leaves the Helper initialization as requiring a `VdafPrepInit` followed by a function similar to `VdafPrepNext` except that the initial `prep_next` call is omitted. (that is, it would call `prep_shares_to_prep` followed by `prep_next`) I think having to specify another arbitrary wrapping function would probably tip this to being a complexity loss -- maybe there's some clever way to cleanly include this case into `VdafPrepNext`?

Reply:
Thinking about this a little more, another approach would to update `VdafPrepInit` to compute `prep_next` conditionally, based on whether it is called by the Leader or the Helper. This function would return different kinds of results based on the role of the caller, but this may still be a relatively-clean way to fix up this issue.

Reply:
The intent is that the Leader can call `VdafPrepInit()`, then immediately call `VdafPrepNext(inbound=None)` to produce the first prep step. The Helper would call `VdafPrepInit()`, then immediately call `VdafPrepNext(inbound=peer_prep_step)`. Then in subsequent rounds, each just calls `VdafPrepNext()`.

So aggregation initialization becomes:
1. Decrypt the report share
1. Do early validation (report replay, etc.)
1. Call `VdafPrepInit()`
2. Call `VdafPrepNext(inbound=None)` if Leader else `VdafPrepNext(inbound=peer_prep_step)`.
3. Send prep step produced from step (2.).

Aggregation continuation becomes:
1. Call `VdafPrepNext(inbound=peer_prep_step)`.
2. Send the prep step produced.

Reply:
Based on in-person discussion, this decomposition works:
* Leader initialization is DapPrepInit -> DapPrepNext -> send/recv -> DapPrepNext.
* Helper initialization is DapPrepInit -> DapPrepNext.
* Leader continuation is send/recv -> DapPrepNext.
* Helper continuation is DapPrepNext.